src/jit/importer.cpp

```diff
@@ -18390,4 +18396,371 @@ bool Compiler::IsMathIntrinsic(GenTreePtr tree)
+        return;
+    }
+
+    JITDUMP("impDevirtualizeCall: type available, attempting devirt\n");
+
+    // Fetch VM handles for base method and base class.
+    CORINFO_METHOD_HANDLE baseMethod = callInfo->hMethod;
+    CORINFO_CLASS_HANDLE  baseClass  = info.compCompHnd->getMethodClass(baseMethod);
+    assert(baseClass != nullptr);
+
+    // In R2R mode, we might see virtual stub calls to
+    // non-virtuals. For instance cases where the non-virtual method
+    // is in a different assembly but is called via CALLVIRT. For
+    // verison resilience we must allow for the fact that the method
+    // might become virtual in some update.
+    //
+    // In non-R2R modes CALLVIRT <nonvirtual> will be turned into a
+    // regular call+nullcheck upstream, so we won't reach this
+    // point.
+    const DWORD baseMethodAttribs = info.compCompHnd->getMethodAttribs(baseMethod);
+    if ((baseMethodAttribs & CORINFO_FLG_VIRTUAL) == 0)
+    {
+        assert(call->IsVirtualStub());
+        assert(opts.IsReadyToRun());
+        JITDUMP("--- [R2R] base method not virtual, sorry\n");
+        return;
+    }
+
+    // If the objClass is sealed (final), then we may be able to devirtualize.
+    const DWORD objClassAttribs = info.compCompHnd->getClassAttribs(objClass);
+    const bool  objClassIsFinal = (objClassAttribs & CORINFO_FLG_FINAL) != 0;
+
+#if defined(DEBUG)
+    const char* const objClassNote   = objClassIsFinal ? " [final]" : "";
+    const char* const objClassName   = info.compCompHnd->getClassName(objClass);
+    const char* const baseClassName  = info.compCompHnd->getClassName(baseClass);
+    const char* const baseMethodName = eeGetMethodName(baseMethod, nullptr);
+
+    if (verbose)
+    {
+        printf("$$$ In %s: Maybe devirt?\n"
+               "    class for 'this' is %s%s (attrib %08x)\n"
+               "    base method is %s::%s\n",
+               info.compFullName, objClassName, objClassNote, objClassAttribs, baseClassName, baseMethodName);
+    }
+#endif // defined(DEBUG)
+
+    // Screen out interface calls.
+    // We might be able to devirtualize these some day.
+    const DWORD baseClassAttribs = info.compCompHnd->getClassAttribs(baseClass);
+    if ((baseClassAttribs & CORINFO_FLG_INTERFACE) != 0)
+    {
+        assert(call->IsVirtualStub());
+        JITDUMP("--- base class is interface, sorry\n");
+        return;
+    }
+
+    // Bail if obj class is an interface.
+    // See for instance System.ValueTuple`8::GetHashCode, where lcl 0 is System.IValueTupleInternal
+    //   IL_021d:  ldloc.0
+    //   IL_021e:  callvirt   instance int32 System.Object::GetHashCode()
+    if ((objClassAttribs & CORINFO_FLG_INTERFACE) != 0)
+    {
+        JITDUMP("--- obj class is interface, sorry\n");
+        return;
+    }
+
+    // Bail if obj class is a shared instance, in particular watch for System::__Canon.
+    // Perhaps overly broad...?
```

Comment:
I believe this will also fire for things like SomeClass<__Canon> which is definitely overly broad.